contrib/openapi.yml

```diff
@@ -0,0 +1,170 @@
+swagger: "2.0"
+info:
+  title: Kostyor API
+  description: Aims & Launches Grenades
+  version: 1.0.0
+schemes:
+  - http
+consumes:
+  - application/json
+produces:
+  - application/json
+paths:
+  /environments:
+    get:
+      summary: List OpenStack Environments
+      responses:
+        200:
+          description: An array of OpenStack environments.
+          schema:
+            type: array
+            items:
+              $ref: '#/definitions/Environment'
+  /environments/{id}:
+    get:
+      summary: Show OpenStack Environment
+      parameters:
+        - name: id
+          type: string
+          pattern: &UUID_PATTERN |
+            ^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$
+          required: true
+          in: path
+          description: |
+            Unique identifier representing the specific OpenStack environment.
+      responses:
+        200:
+          description: An OpenStack environment.
+          schema:
+            $ref: '#/definitions/Environment'
+        404:
+          description: OpenStack environment not found.
+          schema:
+            $ref: '#/definitions/Error'
+  /environments/{id}/upgrade/{action}:
+    put:
+      summary: Control Upgrade Flow of OpenStack Environment
+      parameters:
+        - name: id
+          type: string
+          pattern: *UUID_PATTERN
+          required: true
+          in: path
+          description: |
+            Unique identifier representing the specific OpenStack environment.
+        - name: action
+          type: string
+          enum:
+            - start
+            - pause
+            - abort
+            - continue
+            - rollback
+          required: true
+          in: path
+          description: |
+            An action to be applied for the specified OpenStack environment.
+      responses:
+        200:
+          description: A task instance to track the progress.
+          schema:
+            $ref: '#/definitions/Task'
+        400:
+          description: OpenStack environment is not ready for upgrade.
+          schema:
+            $ref: '#/definitions/Error'
+        404:
+          description: OpenStack environment not found.
+          schema:
+            $ref: '#/definitions/Error'
+  /tasks:
```

Comment:
I thought Kostyor is a domain specific tool, and it won't deal with anything but upgrades. Besides, even if it would do we'll need some sort of deferred tasks and introducing new endpoints to track their progress doesn't seem a good idea in my opinion. So I'd prefer to call it `tasks` and use some `name` attribute to indicate it's an upgrade case, if you insist.